### PR TITLE
Linear solver improvements for Rosenbrock solvers

### PR DESCRIPTION
**Pull request by YingboMa**
@aml5600 could you apply similar changes to other Rosenbrock solvers?

### CODE REVIEW

**Comment by YingboMa** [comment]
Improvements:
1. Only factorize once when W is linear and constant
2. Make Rosenbrock23 work with the recent linear solver interface

File: src/perform_step/rosenbrock_perform_step.jl [hunk 1]
```diff
@@ -37,9 +37,15 @@ end
...
+  new_W = calc_rosenbrock_differentiation!(integrator, cache, γ, γ, repeat_step, false)
+
+  calculate_residuals!(weight, fill!(weight, one(eltype(u))), uprev, uprev,
+                       opts.abstol, opts.reltol, opts.internalnorm, t)
+
+  cache.linsolve(vec(k₁), W, vec(linsolve_tmp), new_W,
+      Pl=DiffEqBase.ScaleVector(weight, true),
+      Pr=DiffEqBase.ScaleVector(weight, false), reltol=opts.reltol)
 
-  cache.linsolve(vec(k₁), W, vec(linsolve_tmp), !repeat_step)
```
Comment (ChrisRackauckas) [inline src/perform_step/rosenbrock_perform_step.jl:42]: do all of the Rosenbrocks need this treatment?

**Comment by ChrisRackauckas** [comment]
The failing test is that the IIP and OOP no longer match outputs now. Are the weights changing the solve at all?

But if it seems right, then this should probably be copied down to all other rosenbrocks.

File: src/perform_step/rosenbrock_perform_step.jl (hunk 1 above)
Comment (YingboMa) [reply src/perform_step/rosenbrock_perform_step.jl:42]: Yes

**Comment by YingboMa** [comment]
The test failure looks unrelated? @ChrisRackauckas

**Comment by ChrisRackauckas** [comment]
https://github.com/SciML/RecursiveArrayTools.jl/issues/137 is the issue.

**Pull request merged**